### ISSUE TITLE
Device and data preparation

### ISSUE THREAD

**Issue by mickahell** [mickahell/qhack21#19]
### Preparing datas
_device.py : includes device circuit functions, e.g. blochhemisphere(theta, phi), classicalstate(bit_array), etc._

- First preparing bunch of parameters and define them as Qats (_north half of a qubit_) or doQs (_south part of a qubit_)
- ex :
<table>
					<thead>
						<tr>
							<th align="center">RX(<i>x</i>)</th>
							<th align="center">RZ(<i>x</i>)</th>
							<th align="center">Qat/doQ</th>
						</tr>
					</thead>
					<tbody>
						<tr>
							<td align="center">π</td>
							<td align="center">0</td>
							<td align="center">doQ</td>
<tr>
<tr>
							<td align="center">π/3</td>
							<td align="center">0</td>
							<td align="center">Qat</td>
						</tr>
<tr>
							<td align="center">2*π/3</td>
							<td align="center">2*π/3</td>
							<td align="center">doQ</td>
						</tr>
</body>
</table>

### Create the device 
- The device who create a qubit at the state `|0>` then will be apply the gate and parameters gave in by the external file
ex : 
<pre>
	|--------------------|
	|       Device       |
	|---------|--|-------|
        | RX(π/3) |--| RZ(0) |----- |S(x)>
	|---------|--|-------|	
	|--------------------|	   
</pre>
- the output of this function will be the qubit in superposition

**Comment by mickahell** [comment]
Idea from @tvarga78 :
<pre>
Device types
------------
1) Unsupervised learning:
   The device outputs quantum states of unknown labels, our task is clustering. We'll see what
   happens using a simulator.
2) Supervised learning:
   We have a labeled training set of device output states. Our task is to train a classifier
   which learns the unknown classification rule. Variants:
   2a) During operation, any given device output can be reproduced, as many times as we wish,
       because we can reproduce the physical circumstances (parameters) that caused the device to
       produce the given output. This would enable us to apply data re-uploading strategies during
       classification.
   2b) During operation, the device outputs cannot be reproduced, because for some reason it's
       not feasible to reproduce the physical circumstances (parameters) mentioned above. So we
       cannot apply data re-uploading strategies. However, during the training session we are in
       full control of the device's parameters in our lab.
</pre>

**Comment by tvarga78** [comment]
So basically, the point in 2a and 2b is whether we can apply data re-uploading or not.

I like the data re-uploading idea a lot: https://arxiv.org/abs/1907.02085

**Comment by mickahell** [comment]
In fact during the training part, you know the data you sent as input so you can produce the same qubit anytime you want

**Comment by tvarga78** [comment]
> In fact during the training part, you know the data you sent as input so you can produce the same qubit anytime you want

Yes, but in 2b, I don't know, the device is take from our nice lab.

**Comment by tvarga78** [comment]
Here is the summary of what we discussed. We target 2 tasks: one is higher priority, the other is lower. The idea is that we make progress with the higher priority task until Wed.

1) **High priority task: 2b**, see picture [here](https://qhack2021.slack.com/files/U01JMUW9TC4/F01ND5PGX8X/plan-exported.002.png).

**Idea:** we classify regions of the Hilbert space. E.g. the Northern hemisphere of the Bloch sphere is "cat", the Southern hemisphere is "dog". As first step, we create a parameterized device that can create state vectors that fall either into the "cat" or "dog" region. Then, we feed the output state of the device into a parameterized classifier circuit (see the VQE 100 exercise for one way how we can make such a circuit). Then, we optimize the classifier same way as in the Circuit Training 500 exercise.

**Catch:** during operation, we cannot calculate expected values, because the device can run only once. So when we calculate the accuracy on the test set, we are not allowed to calculate expected values... we run the circuit just once, make the PauliZ measurement, and if the result is "-1" we say "dog", if it's 1 we say "cat".

**Compare with:** the standard way of testing, that is, when we can run the circuit multiple times to get expected value of the PauliZ measurement. The result can be e.g. 0.5, then we classify it as "cat", as it's closer to 1 than to -1.

**Result to get by Wed:** how much does the accuracy decrease due to the fact that we can run the circuit only once during testing? (Also, can we use a smart cost function during training that will make the accuracy drop less?)

2) **Low priority task: 2a_alt**, see picture [here](https://qhack2021.slack.com/files/U01JMUW9TC4/F01PHPC4EQG/plan-exported.001.png).

TDB

**Comment by Zed-Is-Dead** [comment]
Why not using a general rotation gate U3(theta, phi, lambda) for the state preparation?
I would suggest restricting the rotation angles to a certain region of the Bloch sphere and drawing them at random.
We assume here to deal with pure states - in real life, we would have to consider mixed states/density matrices

**Comment by mickahell** [comment]
Of course @Zed-Is-Dead the schema I wrote is just a suggestion and using a U3 is absolutly good too.

**Comment by Zed-Is-Dead** [comment]
Example of states (100 QCats in red and 100 QDogs in blue) generated at random with some restriction on angles
![image](https://user-images.githubusercontent.com/7906730/108735662-1caed800-7531-11eb-8ca4-e89da1f763f2.png)

**Comment by mickahell** [comment]
@Zed-Is-Dead this is perfect for graphing at the end !

**Comment by mickahell** [comment]
@Zed-Is-Dead I'll update the readme with this image

**Comment by Zed-Is-Dead** [comment]
I have pushed a notebook that generates random qcats and qdogs (cf. picture above).
Also 2 csv files with the cats/dogs angle parameters
Instructions are given at the end on how to recreate the qcats/qdogs from the respective angles

**Comment by mickahell** [comment]
I just used a function in an excel file to generate them in the same time ^^'

**Comment by Zed-Is-Dead** [comment]
I've updated with training/testing data (100 qats + 100 doqs in training and 25/25 for testing)

**Comment by mickahell** [comment]
I pushed a simple U2 sensor, taking in params a simple arrays [x, z]

**Comment by mickahell** [comment]
EDIT, it's now a template circuit who can take an array of 2 or 3 angles (doesn’t matter is the data are 2 or 3 angles)

**Comment by tvarga78** [comment]
**Idea:** we classify regions of the Hilbert space, of quantum states of _n_ qubits. There are 2 categories, "Qat" and "DoQ". E.g. for _n=1_, the Northern hemisphere of the Bloch sphere could be labelled "Qat", the Southern hemisphere "DoQ". The state vectors to classify are generated as the output of a sensor, which is then fed into a classifier circuit of _M_ layers. Note that we are NOT classifying the classical _params_ vector of the sensor, as we could use any other sensor with different parameterization as long as it's capable of producing Qat and DoQ states. Also, we take the sensor as is, we don't try to "optimize" it.

**Catch:** during operation, the sensor can only produce its output once. Thus, when we calculate the accuracy on the test set, we are not allowed to make use of expected values resulting from many shots. There is only 1 shot. (In the training phase, we can optimize using expected values, as training is done in our laboratory... but operation may happen e.g. in a self-driving car.) We'd like to experiment how much the accuracy drops due to this 1-shot limitation, and what kind of cost function would reduce this impact.

![image](https://user-images.githubusercontent.com/72672758/108864637-ac688b00-75f2-11eb-938a-75f7330e3f66.png)

**Extra:** if multiple shots are allowed, how much would a data re-uploading scheme improve the accuracy? E.g. imagine there are _M_ identical sensors located very close to each other. When a certain physical event happens, it sets all the parameters of the _M_ sensors at once, identically. Then, the parameters don't change until the next event. Furthermore, there may be exponentially many parameters of the sensor, inaccessible to us. So again, we are classifying quantum states.

![image](https://user-images.githubusercontent.com/72672758/108864795-d457ee80-75f2-11eb-8929-48ef97000873.png)
